neuralogic/core/constructs/function/__init__.py

```diff
@@ -29,4 +32,78 @@
 Aggregation.SOFTMAX = Softmax("SOFTMAX")
 
 
-__all__ = ["Transformation", "Combination", "Aggregation", "Function"]
+class CombinationWrap:
+    __slots__ = "left", "right", "combination"
+
+    def __init__(self, left, right, combination: Combination):
+        self.left = left
+        self.right = right
+        self.combination = combination
+
+    def __add__(self, other):
+        return CombinationWrap(self, other, Combination.SUM)
+
+    def __mul__(self, other):
+        return CombinationWrap(self, other, Combination.ELPRODUCT)
+
+    def __matmul__(self, other):
+        return CombinationWrap(self, other, Combination.PRODUCT)
+
+    def __str__(self):
+        if not isinstance(self.left, CombinationWrap) and not isinstance(self.right, CombinationWrap):
+            return f"{self.combination}"
+        if not isinstance(self.left, CombinationWrap):
+            return f"{self.combination}({self.right.to_str()})"
+        if not isinstance(self.right, CombinationWrap):
+            return f"{self.combination}({self.left.to_str()})"
+
+        return f"{self.combination}({self.left.to_str()}, {self.right.to_str()})"
+
+    def __iter__(self):
+        if isinstance(self.left, CombinationWrap):
+            for a in self.left:
+                yield a
+        if not isinstance(self.left, CombinationWrap):
+            yield self.left
+
+        if isinstance(self.right, CombinationWrap):
+            for a in self.right:
+                yield a
+        if not isinstance(self.right, CombinationWrap):
+            yield self.right
+
+    def to_combination(self) -> Combination:
+        combination_graph, _ = self._get_combination_node()
+        return MixedCombination(name=self.to_str(), combination_graph=combination_graph)
+
+    def _get_combination_node(self, input_counter: int = 0):
```

Comment:
This whole implementation kinda assumes that the whole body is a mixed combination construct.

RuleBody should:
- Check if there is a CombinationWrap in its body
- Make the whole body into CombinationWrap by using either Combination provided in metadata OR the default one.
- This has to happen before building - after we have metadata ready